protonvpn-cli: 2.2.6 -> 3.11.1, protonvpn-cli_2: init at 2.2.11

WolfangAukang opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This is a major update on ProtonVPN CLI. I was able to test it with my basic account and here are the logs:
```
2022-04-28 03:59:41,850 — cli.py — INFO — __init__:17 — 
-------------------------------------------------

-----------	Initialized protonvpn-cli	-----------

-------------------------------------------------
2022-04-28 03:59:41,850 — cli.py — INFO — __init__:29 — ProtonVPN CLI v3.11.1 (protonvpn-nm-lib v3.8.0; proton-client v0.7.1)
2022-04-28 03:59:41,857 — cli.py — INFO — __init__:64 — CLI command: Namespace(command='c', version=False, help=False, get_logs=False)
2022-04-28 03:59:41,859 — cli.py — INFO — connect:128 — Options: Namespace(servername=None, fastest=False, random=False, cc=None, sc=False, p2p=False, tor=False, protocol=None, help=False)
2022-04-28 03:59:43,968 — cli_dialog.py — INFO — start:59 — Selected country: "United States"
2022-04-28 03:59:48,192 — cli_dialog.py — INFO — start:63 — Selected server: "US-FREE#24"
2022-04-28 03:59:49,087 — cli_dialog.py — INFO — start:67 — Selected protocol: "udp"
2022-04-28 03:59:53,902 — cli_wrapper.py — INFO — _connect:485 — Dbus response: {<ConnectionStartStatusEnum.STATE: 'state'>: <VPNConnectionStateEnum.IS_ACTIVE: 5>, <ConnectionStartStatusEnum.MESSAGE: 'message'>: 'Successfully connected to ProtonVPN.', <ConnectionStartStatusEnum.REASON: 'reason'>: <VPNConnectionReasonEnum.NOT_PROVIDED: 1>}
2022-04-28 03:59:57,697 — cli.py — INFO — __init__:17 — 
-------------------------------------------------

-----------	Initialized protonvpn-cli	-----------

-------------------------------------------------
2022-04-28 03:59:57,697 — cli.py — INFO — __init__:29 — ProtonVPN CLI v3.11.1 (protonvpn-nm-lib v3.8.0; proton-client v0.7.1)
2022-04-28 03:59:57,704 — cli.py — INFO — __init__:64 — CLI command: Namespace(command='d', version=False, help=False, get_logs=False)
2022-04-28 04:00:08,039 — cli.py — INFO — __init__:17 — 
-------------------------------------------------

-----------	Initialized protonvpn-cli	-----------

-------------------------------------------------
2022-04-28 04:00:08,039 — cli.py — INFO — __init__:29 — ProtonVPN CLI v3.11.1 (protonvpn-nm-lib v3.8.0; proton-client v0.7.1)
```

The fix done on `protonvpn-nm-lib` is taken from #129333. Without it, ProtonVPN CLI is not able to find some libraries from NetworkManager

Also, as the previous unofficial version (2) performs the connection through OpenVPN instead of NetworkManager (which does not require root), I am naming it `protonvpn-cli2`. Both are working as intended.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nrdxp commented:
This new versions seems to try to setup a tunnel through network manager. I do use network manager and I have a plus account for protonvpn but I cannot get this to work. That said I do have a semi-special configuration of network manager since it's the only way I've been able reliably avoid DNS leaks while also being able to use dns-over-tls via systemd-resolved:
https://github.com/nrdxp/nrdos/blob/master/profiles/network/networkmanager/default.nix#L4-L8

So perhaps this config values are just causing a problem with the DNS? In any case it hangs when trying to connect.

WolfangAukang commented:
> This new versions seems to try to setup a tunnel through network manager. I do use network manager and I have a plus account for protonvpn but I cannot get this to work.

Yes, I believe this is true because the newest versions (from both GUI and CLI) now use a library called `protonvpn-nm-lib`. Maybe this is the component that is causing conflicts on your case.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/847

SuperSandro2000 commented on pkgs/applications/networking/protonvpn-cli/default.nix:
```diff
@@ -1,29 +1,30 @@
+  propagatedBuildInputs = [
+    protonvpn-nm-lib
+    pythondialog
+  ] ++ [
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/networking/protonvpn-cli/default.nix:
```diff
@@ -1,29 +1,30 @@
 
-python3Packages.buildPythonApplication rec {
+let
+  inherit (python3Packages) buildPythonApplication pythonOlder pythondialog protonvpn-nm-lib;
```
We could use python3Packages.callPackages in top-level to avoid this.

SuperSandro2000 commented on pkgs/applications/networking/protonvpn-cli/2.nix:
```diff
@@ -0,0 +1,41 @@
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Linux command-line client for ProtonVPN (uses openvpn)";
```
```suggestion
    description = "Linux command-line client for ProtonVPN using openvpn";
```

SuperSandro2000 commented on pkgs/applications/networking/protonvpn-cli/2.nix:
```diff
@@ -0,0 +1,41 @@
+  src = fetchFromGitHub {
+    owner = "Rafficer";
+    repo = "linux-cli-community";
+    # There is a tag and branch with the same name. Using commit SHA.
+    rev = "ff12ed8186e63936d36b63ae2bf705c09f27a470";
```
```suggestion
    rev = "refs/tag/${version}";
```

WolfangAukang commented on pkgs/development/python-modules/protonvpn-nm-lib/default.nix:
```diff
@@ -33,7 +35,14 @@ buildPythonPackage rec {
+    })
+  ];
+
+  # Tests can't find systemctl and are not able to access DBus
```
Do you know an example of a package doing this kind of testing? I don't know where to start

WolfangAukang replied:
Another thing to consider: To enable tests for this package besides the DBus issue, we need the `systemd-python` library (currently named `systemd` on [nixpkgs](https://github.com/NixOS/nixpkgs/blob/nixos-21.11/pkgs/development/python-modules/systemd/default.nix#L20)) and also the systemd package, because it looks for `systemctl`.

Is it possible to import both without importing the whole `pkgs` to get `systemd` or would it be the best to rename the Python package into something different (although the problem would be fixing the dependent packages, I guess)?

SuperSandro2000 replied:
e.g. https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/traefik.nix



> Is it possible to import both without importing the whole `pkgs` to get `systemd` or would it be the best to rename the Python package into something different (although the problem would be fixing the dependent packages, I guess)?

You could just input it as a different name 

`` callPackages ../xxx/xxx { python-systemd = systemd; }``

WolfangAukang commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29022,6 +29022,7 @@ with pkgs;
   protonmail-bridge = callPackage ../applications/networking/protonmail-bridge { };
 
   protonvpn-cli = callPackage ../applications/networking/protonvpn-cli { };
+  protonvpn-cli2 = callPackage ../applications/networking/protonvpn-cli/2.nix { };
```
Done

SuperSandro2000 commented on pkgs/development/python-modules/protonvpn-nm-lib/default.nix:
```diff
@@ -33,7 +35,15 @@ buildPythonPackage rec {
+    })
+  ];
+
+  # Checks cannot be run through the sandbox
```
```suggestion
  # Checks cannot be run in the sandbox
```

WolfangAukang committed:
python3Packages.protonvpn-nm-lib: 3.8.0 -> 3.9.0

WolfangAukang committed:
protonvpn-cli: 2.2.6 -> 3.11.1

WolfangAukang committed:
protonvpn-cli_2: init at 2.2.11

Naming as protonvpn-cli_2 as this version uses openvpn to perform connections instead of NetworkManager as on the official version
